src/shogun/base/constraint.h

```diff
@@ -86,18 +91,80 @@ namespace shogun
 	template <typename T>
 	struct generic_checker
 	{
-	public:
-		generic_checker(T val) : m_val(val){};
-		bool operator()(T val) const
+		generic_checker() = default;
+		bool operator()(const T& val) const
 		{
 			return check(val);
 		};
 
 		virtual std::string error_msg() const = 0;
 
+	protected:
+		virtual bool check(const T& val) const = 0;
+	};
+
+	template <typename T>
+	struct custom_constraint: generic_checker<T>
+	{
+		template <typename Functor>
+		custom_constraint(Functor&& func): m_func(func) 
+		{
+		}
+
+		std::string error_msg() const override
+		{
+			return msg;
+		}
+
+	protected:
+		bool check(const T& val) const override 
+		{
+			try
+			{
+				m_func(val);
+			}
+			catch (const std::exception& e)
+			{
+				msg = std::string(e.what());
```

Comment:
it seems like that msg can't be assigned in `check` function, as the `check` is `const `,

Reply:
ah true.. I didn't test this, feel free to fix it :)

Reply:
you can make `msg` `mutable`